added python script for combining git shortlogs across multiple repos#19044

ericluoliu opened the pull request:
https://github.com/zulip/zulip/issues/19044

Running the script through terminal commands

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,83 @@
+            dict[name] = value
+
+def retrieve_log(repo, version):
+    return subprocess.run(['git', 'shortlog', '-s', version], capture_output = True, text = True, 
```
The convention is to not have whitespaces around the `=`, when it is for keyword argument. You should run Black (https://github.com/psf/black) on your repo. It will auto-reformat your code.

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,83 @@
+        cwd = os.path.dirname(pathlib.Path().resolve()) + repo).stdout.splitlines()
+
+def find_version(time, repo):
+    versions_list = subprocess.run(['git', 'tag', '-l'], capture_output = True, text = True, 
```
I usually see `subprocess.check_output` in the repo, e.g. https://github.com/zulip/zulip/blob/master/scripts/lib/setup_venv.py#L181. And so preferrable you should use `check_output` so that the code is uniform.

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,83 @@
+
+def retrieve_log(repo, version):
+    return subprocess.run(['git', 'shortlog', '-s', version], capture_output = True, text = True, 
+        cwd = os.path.dirname(pathlib.Path().resolve()) + repo).stdout.splitlines()
```
This line seems to be repeated about 3 times? You can wrap it in a function.

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,83 @@
+    for version in versions_list:
+        version_time = subprocess.run(['git', 'log', '-1', '--format=%ai', version], capture_output = True, text = True, 
+            cwd = os.path.dirname(pathlib.Path().resolve()) + repo).stdout.split()[0]
+        if(create_Time_Object(version_time) >= create_Time_Object(time)):
```
Not sure. If I were to implement this, I'd find a commit hash of the repo that is closest to the timestamp of the webapp/server Zulip repo, instead of returning the closest tag, which is less precise. Need input from @timabbott.

ericluoliu replied:
OK. I will leave it as is for now.

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,83 @@
+            return version
+    return version
+
+def create_Time_Object(time):
```
I think the idiomatic way to implement this is by using `datetime.strptime`. See e.g. https://stackoverflow.com/questions/466345/converting-string-into-datetime (you don't have to specify the hour and minute).

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,83 @@
+
+time = subprocess.run(['git', 'log', '-1', '--format=%ai', zulip_version], capture_output = True, text = True).stdout.split()[0]
+
+#retrieve versions
```
You can do instead
```python
out_dict = {}  # dict is a reserved keyword, so you shouldn't use it
zulip = retrieve_log("/zulip", zulip_version)
add_log(out_dict, zulip)
for repo_name in ["/zulip-mobile", ...]:
  repo_version = find_version(time, repo_name)
  repo_log = retrieve_log(repo_name, repo_version)
  add_log(out_dict, repo_log)
```

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,83 @@
+time = subprocess.run(['git', 'log', '-1', '--format=%ai', zulip_version], capture_output = True, text = True).stdout.split()[0]
+
+#retrieve versions
+zulip_mobile_version = find_version(time, "/zulip-mobile")
```
The slash prefix makes it look like it is a filesystem path relative to the root. Maybe you should modify `find_version` to automatically add `/`, and so have the argument to be as clean as just "zulip-mobile".

ericluoliu commented:
Made those changes through another commit

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,66 @@
+    return version
+
+
+def path(repo):
```
Function name should be a verb.

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,66 @@
+
+
+# extract git version and time
+if len(sys.argv) > 1:
```
The `zulip_version`, which is the *previous* version for the starting point, should be printed out, in case the script user forgets to specify an argument. So they will know that it defaults to 4.3.

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,66 @@
+if len(sys.argv) > 1:
+    zulip_version = sys.argv[1]
+else:
+    zulip_version = str(4.3)
```
"4.3" is simpler than "str(4.3)"

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,70 @@
+for keys in out_dict:
+    print(str(out_dict[keys]) + "\t" + keys)
+
+print("Zulip Version " + zulip_version)
```
It's rather ambiguous which Zulip version this refers to. Maybe "Previous Zulip version"? The variable `zulip_version` should be renamed to `previous_zulip_version` as well.

rht commented:
@ericluoliu it would be great if you could post an example of the output.

ericluoliu commented:
Running bash command:
$ python3 totalContributors.py 3.0 | sort -nr
Produces the given output

![Screen Shot 2021-07-03 at 8 47 18 PM](https://user-images.githubusercontent.com/54130092/124372586-20214100-dc40-11eb-9999-f9fa2482a6d3.png)

rht commented:
Strange. Why aren't the highest committers not shown first? Should look like this
```
  1017        Anders Kaseorg
   634        Steve Howell
   403        Tim Abbott
   ...
```

rht commented:
Also, I noticed that you didn't specify the version for the latest release, say, `current_version`, and set the timestamp of this version as the boundary for other repos. The fact that it is not specified means that the commits go all the way to today, not the most recent release.

ericluoliu commented:
The output looks correct to me. I only posted the bottom of the output because the list is too big but if I scroll up the top matches similarly to what you have.
![image](https://user-images.githubusercontent.com/54130092/124510542-e27bff80-dd88-11eb-9cb5-003f07c685a2.png)

ericluoliu commented:
Sorry can you explain what you mean by the commits go all the way to today rather than the most recent release. I just set version 4.3 as the default if the user doesn't specify a version so wouldn't the commits go to this version date?

rht commented:
AFAIK "Rishi Gupta" hasn't contributed to any of the repos since 4.3, which means that your shortlog somehow includes commits that are far older than 4.3.

>  I just set version 4.3 as the default if the user doesn't specify a version so wouldn't the commits go to this version date?

If you look at the original command, it specifies a range of versions, not just 1 version:
```
git shortlog -s 3.0..4.0 | sort -nr
```

ericluoliu commented:
Oh, then I must have misunderstood the original issue. I was under the impression that you just wanted a version number and all the contributors from before that version. Can you confirm that my new understanding is correct: we are supposed to have a range of versions e.g. 3.0..4.0 and the contributors list are all the people who contributed within this time frame?

rht commented:
Yes, that's correct.

ericluoliu commented:
Made the changes
python3 totalContributors.py 3.0..4.0 | sort -nr
![image](https://user-images.githubusercontent.com/54130092/124841580-6decc080-df42-11eb-8735-c34b10142bd3.png)
![image](https://user-images.githubusercontent.com/54130092/124841600-7b09af80-df42-11eb-80ad-c2134be92152.png)

The numbers differ slightly because I pull from multiple repos. When the user does not give any arguments I set the default as 1.3.0..4.3

rht commented on totalContributors.py:
```diff
@@ -0,0 +1,83 @@
+            universal_newlines=True,
+            cwd=find_path(repo),
+        ).split()[0]
+        if datetime.strptime(version_time, "%Y-%m-%d") >= datetime.strptime(time, "%Y-%m-%d"):
```
Copy-pasting my previous comment for convenience
> Not sure. If I were to implement this, I'd find a commit hash of the repo that is closest to the timestamp of the webapp/server Zulip repo, instead of returning the closest tag, which is less precise. Need input from @timabbott.

ericluoliu replied:
What are our steps for moving forward with this PR? Do we just wait for now?

rht replied:
Yes, just wait for @timabbott.

MC-Terrence replied:
Hello @ericluoliu  and @rht any feedback on this so far?
I just claimed it and had missed the later part of your conversation.
Let me know if this issue is resolved or still open or pending on @timabbott

ericluoliu replied:
Yes, I think this issue is resolved. Just waiting for @timabbott to provide some input.

MC-Terrence replied:
Ok @ericluoliu . Thank you for the information and well-done man. Good job!
I will rescind my claim now.

timabbott commented:
@ericluoliu can you fix the CI failures in this PR, and write a proper commit message?

It'd also be nice to add tests for it to the `tools/tests` test suite, if we can figure out a reasonable way to do it (Perhaps the tests have a fixture file with some example `git shortlog` output, and can verify the merging logic that way?)

rht commented:
I have been wondering if the script should be put in a separate repo, because it can be used by other projects. But I don't know of any other FOSS projects that list contributions every time they release. This is just a thought. It is fine to just put it in zulip/zulip for now.

ericluoliu commented:
@timabbott I fixed the CI failures and added a [--help] [-h] method but fixing my previous commit messages is giving me a bit of trouble. I am trying to use git rebase -i HEAD~5 to change "pick" to "reword". Supposedly when I run this command, a text document should open with my previous 5 commits but when I run this command, the 5 commits show up in the terminal instead and I do not know how I could change and save that. Any ideas? Thanks!!

rht commented:
Can you take a screenshot of your terminal screen? The default editor is probably `nano`, which always lists the keyboard shortcut in the screen, i.e. ctrl-x to save IIRC.

ericluoliu commented:
![image](https://user-images.githubusercontent.com/54130092/127374027-fd58d191-2b86-455c-b708-8c768ed95400.png)
Here's what it looks like. It gives a list of commands but I am not sure how to use any of them. When I change pick to reword nothing really happens.

rht commented:
What is the output of `echo $EDITOR`? What is your OS?

ericluoliu commented:
Mac OS. echo $EDITOR doesn't give me any output. Or at least it seems empty.

![image](https://user-images.githubusercontent.com/54130092/127552787-575c35f4-e721-4ce0-82d8-0bc51b8cbd95.png)

rht commented:
Those tildes on the screenshot looks like `vim`. If you want to make it default to `nano`, try https://unix.stackexchange.com/questions/501862/how-can-i-set-the-default-editor-as-nano-on-my-mac. Make sure to test it by doing `echo $EDITOR` after restarting your terminal. If you want to stick to vim, you should save-quit by doing `:wq` (you can look this up on the internet).

ericluoliu commented:
Got it, thanks for the pointers! 
I ran into a few problems the other day with the git and the Zulip repository and I ended up deciding to delete my local repository and re-fork it and clone everything again. Do you think it would be possible to create a new PR for this issue to have a clean slate or would it be better to find a way to rename these commits to keep the record?

rht commented:
Making a new PR works, but it is a useful skill to be able to rename commits and squash them as well.

ericluoliu commented:
Yeah, you are right. I am just going to create a new PR for this one though because I no longer have the local-branch "total-contributors" that I was previously working on.
